Fix ng model change

gmazzamuto opened the pull request:
Hi again,
while I was working on the keyboard support I noticed that the `ControlValueAccessor` has some implementation problems. I think that the `ngModelChange` does not need to be called explicitly. Instead,  Angular is notified of changes in the model by simply calling the dedicated function registered through `registerOnChange`. Angular will then take care of firing the `ngModelChange` event for you.

can you please have a look at changes in `src/nouislider.ts` and tell me what you think?

tb commented:
@gmazzamuto This PR changes API - removes `(change)` event (and changes examples). On top of it its breaking Range slider example because onSet seems to be triggered twice.

Check [this](https://github.com/tb/ng2-nouislider/commit/3e5181ce10a4934bf817cd4a9f4eb2345cc8efc1), let me know if it fixes the issues you had.

gmazzamuto commented:
Thanks for your reply, here are a couple of observations.

- You are right that the `(set)` event was called twice; now I fixed it :)
- I didn't remove the `(change)` event, I simply remapped it to nouislider's `change` event

I checked the solution that you proposed. It might work but I think it is not a correct implementation of the `ControlValueAccessor` interface. As far as I understand, the registered callback `onChange()` is not supposed to be called within `writeValue()`. The callback is intended to be used to inform Angular and the outside world that the model has changed internally (for example, when the user interacts with the slider). Instead, the `writeValue()` function serves the opposite function: it is used when Angular wants to write a value into our model.

So when Angular calls the `writeValue()` function, that is not an internal change; rather it's Angular itself that is changing the model, so there's no need to inform about the model change.

what do you think?

tb commented:
@gmazzamuto In your implementation change event is not triggered when I press buttons.

Another thing is you put a lot of logic into `slider.on` `set` which gives inconsistend results like when I move sider I see:

v 7.7
this.value [7.7]
onSet value 7.7
onChange value [7.7]

I need to add more tests so that CI catches this ;-)

gmazzamuto commented:
yes it needs to be refined, I'll let you know in case I come up with something.

As for the `change` event, that's because it is not emitted by nouislider in the first place. Apparently, this is what the developer intended. If you check this table about nouislider's callbacks, it says that the `change` event is not emitted when you press the buttons: [https://refreshless.com/nouislider/events-callbacks/](https://refreshless.com/nouislider/events-callbacks/). There is also a live demo that shows that the `change` event is not triggered.

tb commented:
@gmazzamuto Check https://github.com/tb/ng2-nouislider/pull/22 It includes removal of ngModelChange and update to tests.

Adding 2nd argument to `writeValue` is good approach, but that way I was able to get "most appropriate" `onEvent` calls in tests with least effort so far.

gmazzamuto commented:
a quick update:
- I fixed the issue with the inconsistent data type given by the `set` event
- I also updated the demo. If we bind to `ngModelChange` instead of `change` or `set`, we get the behaviour that you expect (when the slider is clicked and also when the +1 button is pressed).

tb commented:
@gmazzamuto Please merge [add tests](https://github.com/tb/ng2-nouislider/commits/fix_ngModelChange_add_tests).

The `writeValue` could be limited to just 

    writeValue(value: any): void {
      this.slider.set(value);
    }

since all logic is done in `this.slider.on('set', ...)`

Also, this code 

      if(typeof(v) === "number") {
        this.value = v;
      } else {
        this.value = [].concat(v);
      }

is not needed since `map` in `toValue` [already created new array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map). `this.value = v` is fine here. 

Last thing - changing the `(change)` behaviour. My initial idea was to have it working similar to HTML input `(change)`. I would prefer to keep it because it is breaking change to remove it, it will confuse people and does not bring any value. Even the current tests dont have that event triggered atm.

gmazzamuto commented:
@tb Done!

I totally agree, I just removed the unnecessary logic.

I understand your concern with the `(change)` behavior, that's OK.

so everything appears to be in place! :)
